File: 0208-implement-trie-prefix-tree/0208-implement-trie-prefix-tree.cpp
struct Node{
    Node *link[26];
    bool fl=false;
    bool containskey(char ch){
        return link[ch-'a']!=NULL;
    }
    void put(char ch,Node *node){
      link[ch-'a']=node;
    }
    Node* get(char ch){
        return link[ch-'a'];
    }
    void setend(){
        fl=true;
        
    }
    bool isend(){
        return fl;
    }
};
class Trie {
public:
  Node *root;
    Trie() {
        root=new Node();
    }
    
    void insert(string word) {
       Node *node=root;
       for(int i=0;i<word.length();i++){
         if(!node->containskey(word[i])){
          node->put(word[i],new Node());
         }
         node=node->get(word[i]);
       } 
       node->setend();
    }
    
    bool search(string word) {
        Node* node=root;
        for(int i=0;i<word.length();i++){
            if(!node->containskey(word[i])){
                return false;
            }
            node=node->get(word[i]);
        }
        return node->isend();
    }
    
    bool startsWith(string prefix) {
        Node *node=root;
        for(int i=0;i<prefix.length();i++){
            if(!node->containskey(prefix[i])){
                return false;
            }
            node=node->get(prefix[i]);
        }
        return true;
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */

File: 0208-implement-trie-prefix-tree/0208-implement-trie-prefix-tree.py
class TrieNode:
    def __init__(self):
        self.child={}
        self.isend=False
class Trie:

    def __init__(self):
        self.root=TrieNode()
        

    def insert(self, word: str) -> None:
        curr=self.root
        for c in word:
            if c not in curr.child:
                curr.child[c]=TrieNode()
            curr=curr.child[c]
        curr.isend=True
        

    def search(self, word: str) -> bool:
        curr=self.root
        for c in word:
            if c not in curr.child:
                return False
            curr=curr.child[c]
        return curr.isend

        

    def startsWith(self, prefix: str) -> bool:
        curr=self.root
        for c in prefix:
            if c not in curr.child:
                return False
            curr=curr.child[c]
        return True


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)